lightning/src/chain/channelmonitor.rs

```diff
@@ -160,151 +172,202 @@ pub enum MonitorEvent {
 		{
 			let mut monitors = self.monitors.lock().unwrap();
 			for monitor in monitors.values_mut() {
-				let txn_outputs = monitor.block_connected(txn_matched, height, &block_hash, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+				let mut txn_outputs = monitor.block_connected(header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+				has_new_outputs_to_watch |= !txn_outputs.is_empty();
 
-				for (ref txid, ref outputs) in txn_outputs {
-					for (idx, output) in outputs.iter().enumerate() {
-						self.chain_monitor.install_watch_outpoint((txid.clone(), idx as u32), &output.script_pubkey);
+				if let Some(ref chain_source) = self.chain_source {
+					for (txid, outputs) in txn_outputs.drain(..) {
+						for (idx, output) in outputs.iter().enumerate() {
+							chain_source.register_output(&OutPoint { txid, index: idx as u16 }, &output.script_pubkey);
+						}
 					}
 				}
 			}
 		}
+		has_new_outputs_to_watch
 	}
 
-	fn block_disconnected(&self, header: &BlockHeader, disconnected_height: u32) {
-		let block_hash = header.bitcoin_hash();
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel based on the disconnected block. See [`ChannelMonitor::block_disconnected`] for
+	/// details.
+	///
+	/// [`ChannelMonitor::block_disconnected`]: struct.ChannelMonitor.html#method.block_disconnected
+	pub fn block_disconnected(&self, header: &BlockHeader, disconnected_height: u32) {
 		let mut monitors = self.monitors.lock().unwrap();
 		for monitor in monitors.values_mut() {
-			monitor.block_disconnected(disconnected_height, &block_hash, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+			monitor.block_disconnected(header, disconnected_height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
 		}
 	}
 }
 
-impl<Key : Send + cmp::Eq + hash::Hash + 'static, ChanSigner: ChannelKeys, T: Deref, F: Deref, L: Deref, C: Deref> SimpleManyChannelMonitor<Key, ChanSigner, T, F, L, C>
-	where T::Target: BroadcasterInterface,
+impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref, D: Deref> ChainMonitor<ChanSigner, C, T, F, L, D>
+	where C::Target: chain::Notify,
+	      T::Target: BroadcasterInterface,
 	      F::Target: FeeEstimator,
 	      L::Target: Logger,
-        C::Target: ChainWatchInterface,
+	      D::Target: ChannelDataPersister<Keys=ChanSigner>,
 {
 	/// Creates a new object which can be used to monitor several channels given the chain
 	/// interface with which to register to receive notifications.
-	pub fn new(chain_monitor: C, broadcaster: T, logger: L, feeest: F) -> SimpleManyChannelMonitor<Key, ChanSigner, T, F, L, C> {
-		let res = SimpleManyChannelMonitor {
-			monitors: Mutex::new(HashMap::new()),
-			chain_monitor,
+	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F, data_persister: D) -> Result<Self, ChannelMonitorUpdateErr> {
+		let monitors = data_persister.load_channel_data()?;
```

Comment:
Hmm, this is a pretty tricky...*thing*. Monitors may have different current chain tips, and need to have the chain replayed individually for each to get them in sync before we start running proper. I like the idea of doing it in ChainMonitor as a part of the logic, but it needs an API to deal with per-monitor chainstate replay. It may make sense to separate the load_channel_data function out into a separate PR where we can hash out all the details of that.

Reply:
Sg, removed it.